docs/rfc/block-registration.md

```diff
@@ -0,0 +1,441 @@
+* Follow-up issue: [Expose available blocks via an API](https://github.com/WordPress/gutenberg/issues/4116)
+* Current documentation: [https://wordpress.org/gutenberg/handbook/designers-developers/developers/block-api/block-registration/](https://wordpress.org/gutenberg/handbook/designers-developers/developers/block-api/block-registration/)
+* Make WordPress.org post: [The Block Directory, and a new type of plugin](https://make.wordpress.org/meta/2019/03/08/the-block-directory-and-a-new-type-of-plugin/)
+
+
+## Previous attempts
+
+Initial support for server-defined block attributes was merged as part of [#2529](https://github.com/WordPress/gutenberg/pull/2529). PHP block type registrations are merged with those defined in the JavaScript runtime. While this enabled blocks to be defined within PHP, the majority of block types continue to be defined within JavaScript alone. The support was reserved for the exclusive use of [dynamic block types](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/block-tutorial/creating-dynamic-blocks/), in large part because <code>[edit and save behaviors](https://wordpress.org/gutenberg/handbook/designers-developers/developers/block-api/block-edit-save/)</code> must still be implemented in JavaScript, and because a solution hadn't been considered for how to create individual block bundles during the build process, nor how to load such bundles efficiently if it were to come to be implemented.
+
+A demonstration for how block registration could be made filterable in PHP was explored in [#5802](https://github.com/WordPress/gutenberg/pull/5802). The purpose here was to explore how plugins could have better control over the registration.
+
+Another exploration in [#5652](https://github.com/WordPress/gutenberg/pull/5652) considered using JSON as a file format to share block type definitions between JavaScript and PHP. 
+
+### Conclusions
+
+* The current approaches to client-side block type registration cannot support the proposed requirement to have all block types known outside the browser context.
+* Using a statically-defined, JSON-formatted block type definition enables easy integration in both JavaScript and PHP runtimes.
+* Registering a block type in PHP would allow for attribute default values to be assigned as dynamically generated from some external state (e.g. a database value, or localized string).
+* By default, JSON does not support localization or dynamic values.
+* On the server, a block type `icon` property can only be assigned as a string and thus cannot support SVGs and component-based icons.
+
+---
+
+## Introduction
+
+Blocks are the fundamental elements of the editor. They are the primary way in which plugins and themes can register their own functionality and extend the capabilities of the editor.
+
+## Registering a block type
+
+To register a new block type, start by creating a `block.json` file. This file:
```

Comment:
Why would we use a `.json` file instead of following WP conventions for ALL other registries? `register_post_type`, `register_taxonomy`, etc?

Whichever client needs the final Block registry info shouldn't be getting the data from a raw JSON file, but instead should be asking the server (REST Endpoint?) for the Block registry which would pass the registry through filters, translations, etc. 

What value does registering in JSON provide? The client should still expect to get JSON from the server as the final declaration of what a registered block is, but a raw JSON file seems to complicate things. 

For plugin authors that want to register blocks, logic can't be applied to a JSON file to dynamically add more blocks, etc. 

The registry should happen in PHP, so that folks can hook in to filter in additional blocks, use variables, translations, etc within the block registry, etc. 

I think when the client (Gutenberg) asks for the registry, it should be delivered in JSON (REST API, for example) but I don't think .json files make sense as the recommended approach to register a block.

Reply:
Blocks have to be defined in both javascript and php at moment. If we force all blocks to be defined in PHP version, wouldn't this be a breaking change?

Reply:
Not if we just change how the JS client reads the block registry. 

If the JS client read the registry from a REST endpoint or a `wp_localize_script` value, or whatever, it would work fine. 

We'd have to resolve the merger of server side registered blocks and client-registered blocks, but it doesn't have to be a breaking change, as long as the shape of the registry is the same. 

Gutenberg Fields Middleware is a great example of this in action. They've implemented a PHP registry for blocks that plays nice with Gutenberg core. They've gone a step further and even allow you to register fields on your blocks, all from PHP without having to write any custom JS. A Fields API for Gutenberg! With some more expansion, that could be a great approach to the server side registry, and could be fleshed out to support things like server side validation for writing data back to blocks, etc.

Worth checking out if you're not familiar with the project: https://github.com/rtCamp/gutenberg-fields-middleware

Reply:
Another reason to have it as a JSON, so that we have a block registry like plugin repo. This repo could read the json for name and description. How could be do that if dyanimcally generated in PHP?

Reply:
> Blocks have to be defined in both javascript and php at moment. If we force all blocks to be defined in PHP version, wouldn't this be a breaking change?

> Another reason to have it as a JSON, so that we have a block registry like plugin repo. This repo could read the json for name and description. How could be do that if dyanimcally generated in PHP?

@spacedmonkey listed 2 reasons why we JSON format is going to be useful here. Block Directory project is one of the driving factors here. @tellyworth encountered many edge cases when trying to scan the existing plugins to discover what kind of blocks they register.

There is also this `@wordpress/block-editor` subpackage which Gutenberg has which tries to be platform agnostic and with blocks being registered on the client with `registerBlockType` which reads definitions from JSON file it doesn't need to use REST API at all.

Reply:
But a rest api could returned by other platforms. I am not saying not to have the json file, I am saying the javascript could read from the rest api. The benefit of this, is the values are run through php / wordpress and are then filterable / translatable in PHP.

Reply:
I have a draft post for make.wordpress.org/core with the following section for REST API:

> The long term vision for the block discovery in WordPress includes:
> * Fetching the available block types through REST APIs.
> * Fetching block objects from posts through REST APIs.
>
> The proposed implementation for the server-side awareness of block types should ensure that it stays intact with all recommendations that REST API team might have. That's why we encourage the REST API team to get involved in the process early on. There is no immediate need to start working on new block type related API endpoints. However, it would be great to have it included on the roadmap. Ideally, they should stay as close as possible to the WordPress.org API and the final shape of the endpoint for searching for blocks.

So yes, we want to have those block definitions exposed from PHP through REST API or GraphQL in the future. The approach we have as of today won't scale. It allows exposing those definitions through inline JavaScript, and I don't even want to start listing all the drawback as you would probably find a bunch of other reasons why it's a bad idea. I'm sure we are all on the same page with the vision. The only difference is that we need to support also blocks registered only in JavaScript so there needs to be a path for seamless migration from what was shipped in 5.0 to the solution which will fulfill more sophisticated applications.